### PR TITLE
Login

### PR DESCRIPTION
**Pull request by pablo-vazquez01**
all tests and ui is working

### CODE REVIEW

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @IBOutlet weak var warningLabel: UILabel!
+    @IBOutlet weak var submitButton: UIButton!
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        guard checkInput() else { return }
+
+        do {
+            try Credentials.addAcount(username: usernameInput.text ?? "", password: passwordInput.text ?? "")
+            
+        }catch let error as KeychainError {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Brace spacing.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            switch error {
+            case .noPassword:
+                warningLabel.text = "please enter a password"
+            case .unhandledError(status: -25299):
+                warningLabel.text = "That username is already being used. Please try again"
+            case .unexpectdPasswordData:
+                    warningLabel.text = "An error has occured while reading the password. Please try again."
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Indentation is wrong.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            case .unhandledError(let status):
+                warningLabel.text = "An error has occured please try again.  error: \(status)"
+                }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Closing brace looks to be in the wrong place.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            return
+        } catch {
+                warningLabel.text = "Somthing went wrong. Please try again."
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Indentation

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        print("Account Created")
+        warningLabel.text = ""
+        dismiss(animated: true, completion: nil)
+        return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: What is the purpose of this return?

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            case .unexpectdPasswordData:
+                    warningLabel.text = "An error has occured while reading the password. Please try again."
+            case .unhandledError(let status):
+                warningLabel.text = "An error has occured please try again.  error: \(status)"
+                }
+            return
+        } catch {
+                warningLabel.text = "Somthing went wrong. Please try again."
+        }
+        print("Account Created")
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Remove debug print statements.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    
+    func checkInput() -> Bool {
+        let usernameSize = usernameInput.text?.count ?? 0
+        let passwordSize = passwordInput.text?.count ?? 0
+        let passwordConfirmSize = passwordConfirmInput.text?.count ?? 0
+        
+        if usernameSize < 6 {
+            warningLabel.text = "Please enter a username"
+            return false
+        } else if passwordSize < 6  || passwordConfirmSize < 6 {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Do we need to check the password confirm length? It should be covered by password length and the password and confirm password matching.

File: PokemonApp/PokemonApp/ErrorList.swift [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+enum KeychainError: Error {
+    case noPassword
+    case unexpectdPasswordData
+    case unhandledError(status: OSStatus)
+}
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ErrorList.swift]: Leave a blank line at end of file.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            let passwordResult = String(data: passwordData, encoding: String.Encoding.utf8)
+            else {
+                throw KeychainError.unexpectdPasswordData
+        }
+        
+        if password == passwordResult {
+            return true
+        }
+        
+        return false
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Just do a straight return
Instead of
if someBool {
return true
}
return false

do 
return someBool

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+struct Credentials {
+    let username: String
+    let password: String
+    
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Extra blank line. Should only have 1 blank line maximum.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 11]
```diff
@@ -9,12 +9,54 @@
 import UIKit
 
 class ViewController: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Rename ViewController to reflect what the view controller contains.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 12]
```diff
@@ -9,12 +9,54 @@
...
+                loginWarningLabel.text = "An error has occured. please try again"
+            case .unhandledError(let status):
+                loginWarningLabel.text = "An error has occured please try again. error: \(status)"
+                }
+                return
+            } catch {
+                loginWarningLabel.text = "Something went wrong. Please try again"
+            }
+        
+        return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Purpose of this return?

File: PokemonApp/PokemonApp/ViewController.swift [hunk 13]
```diff
@@ -9,12 +9,54 @@
...
+        } catch let error as KeychainError {
+            switch error {
+            case .noPassword:
+                loginWarningLabel.text = "please enter a password"
+            case .unexpectdPasswordData:
+                loginWarningLabel.text = "An error has occured. please try again"
+            case .unhandledError(let status):
+                loginWarningLabel.text = "An error has occured please try again. error: \(status)"
+                }
+                return
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: No need for return here.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 14]
```diff
@@ -9,12 +9,54 @@
...
+            }
+        } catch let error as KeychainError {
+            switch error {
+            case .noPassword:
+                loginWarningLabel.text = "please enter a password"
+            case .unexpectdPasswordData:
+                loginWarningLabel.text = "An error has occured. please try again"
+            case .unhandledError(let status):
+                loginWarningLabel.text = "An error has occured please try again. error: \(status)"
+                }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Indentation looks off here.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 15]
```diff
@@ -9,12 +9,54 @@
...
+        
+        
 
-  override func viewDidLoad() {
-    super.viewDidLoad()
-    // Do any additional setup after loading the view.
-  }
-
-
+    func checkLoginInput() -> Bool {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Should probably be private.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+                return
+        }
+
+        XCTAssertEqual(password, returnPassword)
+        
+        // delete from keychain for retestability
+        var queryDelete = [String : AnyObject]()
+        queryDelete[kSecClass as String] = kSecClassGenericPassword
+        
+        SecItemDelete(queryDelete as CFDictionary)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: This seems like something that should be done in tear down

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+            let returnPassword = String(data: passwordData, encoding: String.Encoding.utf8) else {
+                //if this is entered something is wrong and the test should fail
+                XCTFail()
+                return
+        }
+
+        XCTAssertEqual(password, returnPassword)
+        
+        // delete from keychain for retestability
+        var queryDelete = [String : AnyObject]()
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Dictionaries should be defined like this
[KeyType: ValueType]
Note no space between key and colon.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+       let query = setupQuery()
+        
+        var queryResult: AnyObject?
+        withUnsafeMutablePointer(to: &queryResult) {
+            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+        }
+
+        guard let existingItem = queryResult as? [String : AnyObject],
+            let passwordData = existingItem[kSecValueData as String] as? Data,
+            let returnPassword = String(data: passwordData, encoding: String.Encoding.utf8) else {
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: There should probably be a function that gets the password for you.

File: PokemonApp/PokemonApp/ErrorList.swift (hunk 8 above)
Comment (prrenfrew) [reply PokemonApp/PokemonApp/ErrorList.swift]: Still needs the extra line.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        guard let existingItem = queryResult as? [String : AnyObject],
+            let passwordData = existingItem[kSecValueData as String] as? Data,
+            let passwordResult = String(data: passwordData, encoding: String.Encoding.utf8)
+            else {
+                throw KeychainError.unexpectdPasswordData
+        }
+        
+        return password == passwordResult 
+    }
+    private static func keychainQuery( username: String?, accessGroup: String? = nil) -> [String : AnyObject] {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: There should be a line between functions.

File: PokemonApp/PokemonApp/LoginView.swift [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+//  ViewController.swift
+//  PokemonApp
+//
+//  Created by MAC on 6/6/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import UIKit
+
+class LoginController: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: Final class

File: PokemonApp/PokemonApp/LoginView.swift [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+            switch error {
+            case .noPassword:
+                loginWarningLabel.text = "please enter a password"
+            case .unexpectdPasswordData:
+                loginWarningLabel.text = "An error has occured. please try again"
+            case .unhandledError(let status):
+                loginWarningLabel.text = "An error has occured please try again. error: \(status)"
+            }
+        } catch {
+                loginWarningLabel.text = "Something went wrong. Please try again"
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: Indentation looks incorrect.

File: PokemonApp/PokemonApp/UserCredentials.swift [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+struct Credentials {
+    let username: String
+    let password: String
+    
+    init(username: String, password: String) {
+        self.username = username
+        self.password = password
+    }
+}
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/UserCredentials.swift]: Blank line at EOF.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 23]
```diff
@@ -8,13 +8,50 @@
...
 import UIKit
 
-class ViewController: UIViewController {
-
-  override func viewDidLoad() {
-    super.viewDidLoad()
-    // Do any additional setup after loading the view.
-  }
-
+class loginView: UIViewController {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Final class

Comment (prrenfrew) [reply PokemonApp/PokemonApp/ViewController.swift]: Also, use upper camel case LikeThisForClasses

File: PokemonApp/PokemonApp/ViewController.swift [hunk 24]
```diff
@@ -8,13 +8,50 @@
...
+    
+    @IBOutlet weak var usernameTextfield: UITextField!
+    @IBOutlet weak var passworldTextfield: UITextField!
+    @IBOutlet weak var loginButton: UIButton!
+    @IBOutlet weak var newAccountButton: UIButton!
+    @IBOutlet weak var loginWarningLabel: UILabel!
+    
+    @IBAction func loginClicked(_ sender: Any) {
+        guard checkLoginInput() else { return }
+        do {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: This looks like business logic. Put it in the controller layer.

File: PokemonApp/PokemonApp/ViewController.swift [hunk 25]
```diff
@@ -8,13 +8,50 @@
...
+            switch error {
+            case .noPassword:
+                loginWarningLabel.text = "please enter a password"
+            case .unexpectdPasswordData:
+                loginWarningLabel.text = "An error has occured. please try again"
+            case .unhandledError(let status):
+                loginWarningLabel.text = "An error has occured please try again. error: \(status)"
+            }
+        } catch {
+                loginWarningLabel.text = "Something went wrong. Please try again"
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/ViewController.swift]: Indentation

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import XCTest
+import Foundation
+@testable import PokemonApp
+
+final class PokemonAppTests: XCTestCase {
+
+    private var username = "tester1"
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Why are these vars instead of lets?

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+    private var password = "testerspassword1"
+    private var username2 = "tester2"
+    private var password2 = "testerpassword2"
+    
+    override func setUp() {
+        try! Credentials.addAcount(username: username, password: password)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Call super.setup() before your setup.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    
+    override func tearDown() {
+        var queryDelete = [String:AnyObject]()
+        queryDelete[kSecClass as String] = kSecClassGenericPassword
+        
+        SecItemDelete(queryDelete as CFDictionary)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Call super.tearDown() after your tear down logic.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+
+        
+        let returnPassword = getPassword(user: username2)
+
+        XCTAssertEqual(password2, returnPassword)
+    }
+    
+    func testLogin(){
+
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Too many blank lines.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+    
+    @IBAction func clickedSubmit(_ sender: Any) {
+        guard checkInput() else { return }
+
+        do {
+            try Credentials.addAcount(username: usernameInput.text ?? "", password: passwordInput.text ?? "")
+            
+        } catch let error as KeychainError {
+            switch error {
+            case .noPassword:
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: This looks like business logic. Create a function that returns a string that does this checking.

File: PokemonApp/PokemonApp/CreateAccountViewController.swift [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+        if usernameSize < 6 {
+            warningLabel.text = "Please enter a username"
+            return false
+        } else if passwordSize < 6 {
+            warningLabel.text = "Please enter a password"
+            return false
+        } else if passwordInput.text != passwordConfirmInput.text {
+            warningLabel.text = "Passwords must match"
+            return false
+        }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/CreateAccountViewController.swift]: Do this in controller

File: PokemonApp/PokemonApp/LoginView.swift [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+            if success {
+                    //login alert
+                loginWarningLabel.text = "login successful"
+                
+            } else {
+                loginWarningLabel.text = "Incorrect username or password. Please try again"
+            }
+        } catch let error as KeychainError {
+            switch error {
+            case .noPassword:
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: Put business logic in controller.

File: PokemonApp/PokemonApp/LoginView.swift [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+        } catch {
+            loginWarningLabel.text = "Something went wrong. Please try again"
+        }
+    }
+
+    private func checkLoginInput() -> Bool {
+        let usernameSize = usernameTextfield.text?.count ?? 0
+        let passwordSize = passworldTextfield.text?.count ?? 0
+        
+        if usernameSize < 6 {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginView.swift]: Put business logic in controller.

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 34]
```diff
@@ -0,0 +1,82 @@
...
+    
+    private func setupQuery(use: String) -> [String: AnyObject] {
+        
+        var query = [String : AnyObject]()
+        query[kSecClass as String] = kSecClassGenericPassword
+        query[kSecAttrAccount as String] = use as AnyObject?
+        query[kSecMatchLimit as String] = kSecMatchLimitOne
+        query[kSecReturnAttributes as String] = kCFBooleanTrue
+        query[kSecReturnData as String] = kCFBooleanTrue
+
```
Comment (jalwin) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: take off space

File: PokemonApp/PokemonApp.xcodeproj/project.pbxproj [hunk 35]
```diff
@@ -16,8 +16,9 @@
 		029E74FD22AE030C004C09BC /* CreateAccountViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 029E74FC22AE030C004C09BC /* CreateAccountViewController.swift */; };
 		029E74FF22B14B74004C09BC /* ErrorList.swift in Sources */ = {isa = PBXBuildFile; fileRef = 029E74FE22B14B74004C09BC /* ErrorList.swift */; };
 		029E752022BAD87F004C09BC /* NewAccountTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 029E751F22BAD87F004C09BC /* NewAccountTests.swift */; };
+		02F33F9B22BD5CAA00E23156 /* loginTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 02F33F9A22BD5CAA00E23156 /* loginTests.swift */; };
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp.xcodeproj/project.pbxproj]: File names should be in UpperCamelCase.

File: PokemonApp/PokemonApp.xcodeproj/project.pbxproj [hunk 36]
```diff
@@ -121,6 +123,7 @@
 				02307EBD22B05B1E004C34BB /* PokemonAppTests.swift */,
 				02307EBF22B05B1E004C34BB /* Info.plist */,
 				029E751F22BAD87F004C09BC /* NewAccountTests.swift */,
+				02F33F9A22BD5CAA00E23156 /* loginTests.swift */,
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp.xcodeproj/project.pbxproj]: File names should be in UpperCamelCase.

File: PokemonApp/PokemonApp.xcodeproj/project.pbxproj [hunk 37]
```diff
@@ -322,16 +325,17 @@
 				029DC0C122B3E1FE00E7079F /* URLMockSessionInvalid.swift in Sources */,
 				029E752022BAD87F004C09BC /* NewAccountTests.swift in Sources */,
 				02307EE122B12725004C34BB /* PokemonAppTests.swift in Sources */,
+				02F33F9B22BD5CAA00E23156 /* loginTests.swift in Sources */,
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp.xcodeproj/project.pbxproj]: UpperCamelCase

File: PokemonApp/PokemonApp/LoginController.swift [hunk 38]
```diff
@@ -24,6 +24,26 @@ final class KeychainActions {
...
 
+    static func login(username: String, password: String) throws -> Bool {
+        let query = keychainQuery(username: username)
+        
+        var queryResult: AnyObject?
+        let status = withUnsafeMutablePointer(to: &queryResult){
+            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+        }
+        
+        guard status != errSecItemNotFound else { throw KeychainError.incorrectPassword}
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: There needs to be a space before `}`

File: PokemonApp/PokemonApp/LoginController.swift [hunk 39]
```diff
@@ -24,6 +24,26 @@ final class KeychainActions {
...
+        guard status == noErr else { throw KeychainError.unhandledError(status: status)}
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: There needs to be a space before }.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 40]
```diff
@@ -48,10 +67,11 @@ final class KeychainActions {
             let created = try KeychainActions.addAcount(account: username, password: password)
             if created {
                 return "Account created"
-                
             }
         } catch let error as KeychainError {
             switch error {
+            case .incorrectPassword:
+                return "u wot m8"
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Funny, but we should have a bit more descriptive message.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 41]
```diff
@@ -76,5 +96,42 @@ final class KeychainActions {
...
+    static func doLogin(usernameTextfield: String, passwordTextfield: String) -> String{
+        guard checkLoginInput(usernameSize: usernameTextfield.count, passwordSize: passwordTextfield.count) else { return "Incorrect username or password. Please try again." }
+        do {
+            let success = try KeychainActions.login(username: usernameTextfield, password: passwordTextfield)
+            if success {
+                //login alert
+                return "login successful"
+            } else {
+                return "Incorrect username or password. Please try again."
+            }
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: This is a good use case for ternary.
`return success ? trueMessage : falseMessage`

File: PokemonApp/PokemonApp/LoginController.swift [hunk 42]
```diff
@@ -76,5 +96,42 @@ final class KeychainActions {
...
+        } catch {
+            return "Something went wrong. Please try again"
+        }
+    }
+
+    static private func checkLoginInput(usernameSize: Int, passwordSize: Int) -> Bool {
+        
+        if usernameSize < 6 {
+            return false
+        } else if passwordSize < 6 {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Combine these two if's together with `||`

File: PokemonApp/PokemonAppTests/PokemonAppTests.swift [hunk 43]
```diff
@@ -81,5 +80,6 @@ final class PokemonAppTests: XCTestCase {
         }
         waitForExpectations(timeout: 0.1, handler: nil)
     }
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonAppTests.swift]: Why was this line added?

File: PokemonApp/PokemonAppTests/loginTests.swift [hunk 44]
```diff
@@ -0,0 +1,43 @@
...
+//  PokemonAppTests
+//
+//  Created by Mac Consultant on 6/19/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import XCTest
+@testable import PokemonApp
+
+class loginTests: XCTestCase {
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/loginTests.swift]: UpperCamelCase. Also final class.

File: PokemonApp/PokemonAppTests/loginTests.swift [hunk 45]
```diff
@@ -0,0 +1,43 @@
+//
+//  loginTests.swift
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/loginTests.swift]: Update the file name up in the comments.

File: PokemonApp/PokemonApp/LoginController.swift [hunk 46]
```diff
@@ -76,5 +96,34 @@ final class KeychainActions {
         }
         return true
     }
+
+
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/LoginController.swift]: Too many blank lines.
